### DISCUSSION TITLE
Can I use conditional encoder configurations ?

### DISCUSSION THREAD

**Discussion by Russsgithub** [savonet/liquidsoap#1915] [Q&A]
While using %ffmpeg as an encoder for output.icecast with a switch, i'd like to be able to change from %audio.copy to %audio(codec="libmp3lame", samplerate="44100", bitrate="320k") depending on a switch. The switch changes the stream from input.http ( which doesn't need encoding ) to a playlist which does need encoding.

I've tried this
```
pl = playlist('./playlist)
#### buffer for use in multiple places
pl = buffer(mksafe(pl))

s = input.http(.....) 
t = interactive.bool("switch")

s = switch([(t,s),({true}, pl)])

def enc()
  if t then
    (%ffmpeg(format="fdkaac", %audio.copy))
  else
    (%ffmpeg(format="mp3", %audio(codec = "libmp3lame", samplerate=44100,b="320k")))
  end
end
encoder = enc()
s = output.icecast(encoder, fallible = true, host="#{!out_url}", port=8000, password="#{!out_pass}", mount="#{!out_mount}", s)
```

but get this error
```
Error 5: this value has type
  (...) -> format(audio=ffmpeg.audio.copy(_),...)
but it should be a subtype of
  (...) -> format(audio=pcm(_),...)?
```
at the if t then line.

I've noticed there are issues with using %ffmpeg with mksafe(s) before it, but if I remove ```buffer(mksafe(s))``` I get this error
```
Error 5: this value has type
  _ * source(audio=?A, video=?B, midi=?C)
  .{
    time : () -> float,
    shutdown : () -> unit,
    fallible : bool,
    skip : () -> unit,
    seek : (float) -> float,
    is_active : () -> bool,
    is_up : () -> bool,
    log : 
    {level : (() -> int?).{set : ((int) -> unit)}
    },
    self_sync : () -> bool,
    duration : () -> float,
    elapsed : () -> float,
    remaining : () -> float,
    on_track : ((([string * string]) -> unit)) -> unit,
    on_leave : ((() -> unit)) -> unit,
    on_shutdown : ((() -> unit)) -> unit,
    on_metadata : ((([string * string]) -> unit)) -> unit,
    is_ready : () -> bool,
    id : () -> string,
    buffer_length : () -> float,
    status : () -> string,
    set_url : ({string}) -> unit,
    url : () -> string,
    stop : () -> unit,
    start : () -> unit,
    is_started : (() -> bool)
  }
but it should be a subtype of the type of the value at HttpStreamMixerv2_ffmpeg_pass_through.liq, line 216, char 52-73
  _ * _.{length : _}
```
at the switch line.

Thanks

**Comment by toots** [comment]
This is a feature that is being developed by @smimram as part of the `2.1.x` branch. You might want to try the latest `main` to test it!

**Comment by Russsgithub** [comment]
@toots
@smimram 
I thought I was, but have noticed some updates since I tried , so i'll give it a go.

Still getting errors
```
Error 5: this value has type
  (...) -> format(audio=ffmpeg.audio.copy(_),...)
but it should be a subtype of
  (...) -> format(audio=pcm(_),...)?
  ```
  but it could be my script. I'll look into it further later.
  
  Thanks

**Reply by smimram** [reply]
My patch is supposed to help you dynamically change the encoder, but here it won't work: you can only switch between encoders with the same type, but here you are precisely trying to chanegen between encoded and not... We would need to allow something like `audio=any` as type which states that we can have audio which is anything...

**Reply by Russsgithub** [reply]
@smimram 
If you up for doing it , i'm for for trying to use it.

**Reply by chrismahoney** [reply]
Also interested in this if it relates to some things I’m looking into now that I’m up on 2.0 rc1

**Reply by smimram** [reply]
Oh wait, this is more complicated than just this. I read this quickly and did not notice that you want to switch between the encoders *at runtime*. This will not work easily even if we fix the typing system. We need to discuss this with @toots first: do you think that we could ever switch between internal and encoded frames?

**Reply by Russsgithub** [reply]
In the case I opened the discussion for, I'm creating the above as a dynamic source, which is not active at creation. I'm thinking at the moment of catching the 404 error from the input until the source comes up, then trying a pass through encode (audio.copy) , and if this fails changing the encoder to mp3/acc. This would enable me to save CPU usage if the incoming stream fits the pass through chain criteria, in my case aac 48k 24bit , and if not encode it suitably.

**Reply by Russsgithub** [reply]
I'd have to drop the mksafe for a silence file fallback for audio.copy also, I noticed

**Reply by toots** [reply]
That's a very interesting idea. I think the first thing we should consider here is a pro-active operator that allows you to test compatibility and particularly with ffmpeg copy encoder without having to setup the whole streaming machinery before catching it up.

Would you mind sharing a quick example of your code so I can get an idea about how you're going at it?

**Reply by smimram** [reply]
Any chance you could try #1946?

**Reply by Russsgithub** [reply]
@smimram 

I've just tried this
```
def enc()
  if true == true then
    %ffmpeg(format="adts", %audio.copy)
  else
    %ffmpeg(format="adts", %audio(%fdkaac(channels=2, samplerate=48000, bitrate=320, afterburner=true, transmux="adts", aot="mpeg4_aac_lc", sbr_mode=false))) 
  end
end

enc = enc()

output.icecast(fallible=true, enc, format="audio/aac", host="HOST", password="PASS", port=8000, mount="MOUNT", s)
```
with the latest git main code, but got this error
```
Error 5: this value has type
  (...) -> format(audio=pcm(_),...)
but it should be a subtype of
  (...) -> format(audio=ffmpeg.audio.copy(_),...)
```
Thanks

**Reply by smimram** [reply]
Did you really try on the `source-any` branch? Here, your example is going through and I get the "expected" types:
```
enc   : () -> format(audio=any(), video=none, midi=none)
enc   : format(audio=any(), video=none, midi=none)
```

**Reply by Russsgithub** [reply]
@smimram 
No. sorry I misunderstood your post, i'll try it now. Thanks

**Reply by Russsgithub** [reply]
Yep, apologies again for the mistake. This is working as expected , thanks.

**Comment by Russsgithub** [comment]
Will do in a couple of hours. Thanks

On 17 Sep 2021 14:08, Romain Beauxis ***@***.***> wrote:

That's a very interesting idea. I think the first thing we should consider here is a pro-active operator that allows you to test compatibility and particularly with ffmpeg copy encoder without having to setup the whole streaming machinery and catching it up.

Would you mind sharing a quick example of your code so I can get an idea about how you're going at it?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/savonet/liquidsoap/discussions/1915#discussioncomment-1347453>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB627QOSXBFASUWHACC72XLUCM4T7ANCNFSM5D6CV2TA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by Russsgithub** [comment]
@toots 

I'm writting this script for someone else but have there permission to share code, so I'll get a script together , i need to have a think.

Thanks

**Comment by toots** [comment]
I was looking at your script in relation to #19413 and had a couple of thoughts about how to do it with the current `main` branch.

First, I totally understand the need to relax the typing system as @smimram is trying to do, this'll make everything much easier to write. However, it's not clear if we can do it safely or, at least, this might need some time to be implemented right.

Meanwhile, here's how you can do it with the current typing system:
1. Encapsulate all sources + encoders operations in a single function
2. Use closures to store in list of existing sources

For 1. this means cramming everything that uses specific encoder type into a single function.

For 2. this means asking yourself what you really need to _do_ on the created source and store that instead of the source itself.

Something like this:
```ruby
def create_source_with_encoder(encoder, params) =
  s = single("/path/to/security.m4a")

  ....

  output.icecast(encoder, s, ...)
  
  def shutdown() =
    source.shutdown(s)
  end
  
  shutdown_callbacks := [( id, shutdown), ...!shutdown_callbacks]
end

if should_encode then
  create_source_with_encoder(%ffmpeg(format="adts", %audio(codec="aac"), ...), ...)
else
  create_source_with_encoder(%ffmpeg(format="adts", %audio.copy, ...), ...)
end  

def shutdown_source(id) =
  shutdown = list.assoc(id, !shutdown_callbacks)
  shutdown()
end
```
This way, you encapsulate the type of the encoder inside `create_source_with_encoder`, which becomes of type: `(encoder('a), ...) -> unit` and, similarly, your list of existing sources is just a list of shutdown callbacks of type `() -> unit` and you can live in peace mixing different encoders.

Hope this makes sense!

**Reply by Russsgithub** [reply]
Thanks.

I've tried to integrate this code with mine, but get syntax error at the end of shutdown on this line

```
shutdown_callbacks := (id, shutdown)::!shutdown_callbacks
```

To be honest I don't understand the double colon.

I have shutdown_callbacks initiated as ref([]),(i guess it is a list), id as a tuple (x,y) and shutdown as in your script.

**Reply by toots** [reply]
> Thanks.
> 
> I've tried to integrate this code with mine, but get syntax error at the end of shutdown on this line
> 
> ```
> shutdown_callbacks := (id, shutdown)::!shutdown_callbacks
> ```
> 
> To be honest I don't understand the double colon.
> 
> I have shutdown_callbacks initiated as ref([]),(i guess it is a list), id as a tuple (x,y) and shutdown as in your script.

Oh my, sorry that was me mixing OCaml and liquidsoap syntax..

My example wasn't meant to be directly integrated to your script, just a quick schematic example of the ideas I was exposing.

I'd say we should first focus on confirming that the segfault with ffmpeg is fixed then we can play around cleaning up the script.. 🙂

**Reply by Russsgithub** [reply]
I was more integrating your ideas into my script.

I managed to get it running by replacing the ocaml code with ```list.add((id, shutdown),!shutdown_callbacks)``` but still get the encoder mismatch error ```got %ffmpeg(audio.copy) should be internal(a)```

No stress though.

**Comment by Russsgithub** [comment]
Seg fault seems to be fixed, ran a 20 min test that didn't crash.

On 21 Sep 2021 14:45, Romain Beauxis ***@***.***> wrote:

Thanks.

I've tried to integrate this code with mine, but get syntax error at the end of shutdown on this line

shutdown_callbacks := (id, shutdown)::!shutdown_callbacks



To be honest I don't understand the double colon.

I have shutdown_callbacks initiated as ref([]),(i guess it is a list), id as a tuple (x,y) and shutdown as in your script.

Oh my, sorry that was me mixing OCaml and liquidsoap syntax..

I'd say we should first focus on confirming that the segfault with ffmpeg is fixed then we can play around cleaning up the script.. 🙂

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/savonet/liquidsoap/discussions/1915#discussioncomment-1362399>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB627QJ6PEKLFG3B5DN6AFLUDCD7BANCNFSM5D6CV2TA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by toots** [reply]
Wonderful!
